proto/frontend/src/common.cpp

```diff
@@ -99,6 +99,44 @@ std::string bytestring_pi_to_p4rt(const char *str, size_t n) {
   return std::string(str + i, n - i);
 }
 
+Status bytestring_to_pi_port(const std::string &str, pi_port_t* result) {
+  if (str.empty()) {
+    RETURN_ERROR_STATUS(Code::INVALID_ARGUMENT,"Port must not be the empty string");
+  }
+  if (str.size() > 4) {
+    RETURN_ERROR_STATUS(Code::UNIMPLEMENTED,
+                        "Got port of %d bytes, but only 4 bytes are supported",
+                        str.size());
+  }
+
+  uint32_t value = 0;
+  for (char byte : str) {
+    value <<= 8;
+    value += static_cast<uint8_t>(byte);
+  }
+  // Casting from unsigned to signed is implementation-defined if the source
+  // value would not fit in the destination, so we use memcpy as a work-around.
+  static_assert(sizeof(pi_port_t) == 4, "invariant broken");
+  memcpy(result, &value, 4);
+
+  if (*result <= 0) {
+    RETURN_ERROR_STATUS(Code::INVALID_ARGUMENT,"Port must be non-negative.");
+  }
```

Comment:
so you don't think we need to check for `SDN_PORT_UNKNOWN` here, or at the caller?

Reply:
`SDN_PORT_UNKNOWN == 0`, so technically we are checking.

If we are okay with pulling in the P4RT dependency here, I can make this explicit?

Reply:
current code lgtm

Reply:
SG